File: WebClient SDK/src/default/workflow/krProcess/krTileInflater.ts
import { IKrTileCommand } from './interfaces';
import { KrGlobalTileCommand } from './krGlobalTileCommand';
import { KrLocalTileCommand } from './krLocalTileCommand';
import {
  TileContextSource,
  ITile,
  Tile,
  TileEvaluationEventArgs,
  TileGroups
} from 'tessa/ui/tiles';
import { KrTileInfo } from 'tessa/workflow/krProcess';
import { IStorage } from 'tessa/platform/storage';
import { tryGetFromInfo, showNotEmpty } from 'tessa/ui';
import { getTessaIcon } from 'common/utility';
import { ValidationResult } from 'tessa/platform/validation';

export class KrTileInflater {
  //#region ctor

  private constructor() {}

  //#endregion

  //#region instance

  private static _instance: KrTileInflater;

  public static get instance(): KrTileInflater {
    if (!KrTileInflater._instance) {
      KrTileInflater._instance = new KrTileInflater();
    }
    return KrTileInflater._instance;
  }

  //#endregion

  //#region methods

  public inflate(
    contextSource: TileContextSource,
    tileInfos: ReadonlyArray<KrTileInfo>,
    groups: number | null = null
  ): ITile[] {
    const tiles: ITile[] = [];

    for (const tileInfo of tileInfos) {
      tiles.push(this.inflateTile(contextSource, tileInfo, groups));
    }

    return tiles;
  }

  private inflateTile(
    contextSource: TileContextSource,
    tileInfo: KrTileInfo,
    groups: number | null = null
  ): ITile {
    const icon = getTessaIcon(tileInfo.icon);

    const tileCollection: ITile[] = [];
    if (tileInfo.nestedTiles.length !== 0) {
      const nestedTiles: ITile[] = [];
      for (const nestedTileInfo of tileInfo.nestedTiles) {
        nestedTiles.push(this.inflateTile(contextSource, nestedTileInfo, groups));
      }

      tileCollection.push(...nestedTiles);

      // Тайл группы.
      const grouptile = new Tile({
        name: tileInfo.caption,
        caption: tileInfo.caption,
        icon: icon,
        contextSource,
        order: tileInfo.order,
        group: groups ?? TileGroups.NoGroup,
        tiles: tileCollection,
        sharedInfo: KrTileInflater.createInfo(tileInfo),
        toolTip: tileInfo.tooltip
      });
      if (tileInfo.actionGrouping) {
        grouptile.info['.actionsGrouping'] = true;
        grouptile.evaluating.add(KrTileInflater.tileGroupingEvaluation);
      }
      return grouptile;
    }

    const tile = new Tile({
      name: tileInfo.name,
      caption: tileInfo.caption,
      icon: icon,
      contextSource,
      command: KrTileInflater.onClickAction,
      order: tileInfo.order,
      group: groups ?? TileGroups.NoGroup,
      tiles: tileCollection,
      sharedInfo: KrTileInflater.createInfo(tileInfo),
      toolTip: tileInfo.tooltip
    });
    if (tileInfo.actionGrouping) {
      tile.info['.actionsGrouping'] = true;
      tile.evaluating.add(KrTileInflater.tileGroupingEvaluation);
    }
    return tile;
  }

  private static async onClickAction(tile: ITile) {
    const tileInfo = tryGetFromInfo<KrTileInfo | null>(tile.sharedInfo, 'TileInfo', null);
    if (tileInfo) {
      const context = tile.context;
      const command: IKrTileCommand = tileInfo.isGlobal
        ? KrGlobalTileCommand.instance
        : KrLocalTileCommand.instance;

      try {
        // В ТК вызывается как async void.
        // UIContext не будет сохраняться после первой асинхронной операции.
        // это нужно для синхронизации поведения, например в AdvancedDialogCommandHandler.
        command.onClickAction(context, tile, tileInfo);
      } catch (err) {
        await showNotEmpty(ValidationResult.fromError(err));
      }
    }
  }

  private static createInfo(tileInfo: KrTileInfo): IStorage {
    return {
      TileInfo: tileInfo
    };
  }

  private static tileGroupingEvaluation(e: TileEvaluationEventArgs) {
    e.setIsEnabledWithCollapsing(e.currentTile, true);
  }

  //#endregion
}

File: Source/Extensions/Tessa.Extensions.Default.Client/Workflow/KrProcess/KrTileInflater.cs
using System.Collections.Generic;
using System.Linq;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.Platform.Collections;
using Tessa.Platform.Storage;
using Tessa.UI;
using Tessa.UI.Tiles;
using Unity;

namespace Tessa.Extensions.Default.Client.Workflow.KrProcess
{
    public sealed class KrTileInflater : IKrTileInflater
    {
        private readonly IIconContainer iconContainer;

        private readonly IKrTileCommand localCommand;

        private readonly IKrTileCommand globalCommand;

        public KrTileInflater(
            IIconContainer iconContainer,
            [Dependency(KrTileCommandNames.Local)] IKrTileCommand localCommand,
            [Dependency(KrTileCommandNames.Global)] IKrTileCommand globalCommand)
        {
            this.iconContainer = iconContainer;
            this.localCommand = localCommand;
            this.globalCommand = globalCommand;
        }

        /// <inheritdoc />
        public List<ITile> Inflate(
            ITileContextSource contextSource,
            IReadOnlyCollection<KrTileInfo> tileInfos)
        {
            var tiles = new List<ITile>(tileInfos.Count);

            foreach (var tileInfo in tileInfos)
            {
                tiles.Add(this.InflateTile(contextSource, tileInfo));
            }

            return tiles;
        }

        private ITile InflateTile(
            ITileContextSource contextSource,
            KrTileInfo tileInfo)
        {
            IIcon icon;
            try
            {
                icon = this.iconContainer.Get(tileInfo.Icon);
            }
            catch (KeyNotFoundException)
            {
                icon = Icon.Empty;
            }

            var tileCollection = new TileCollection();
            if (tileInfo.NestedTiles.Count != 0)
            {
                var nestedTiles = new List<ITile>(tileInfo.NestedTiles.Count);
                foreach (var nestedTileInfo in tileInfo.NestedTiles)
                {
                    nestedTiles.Add(this.InflateTile(contextSource, nestedTileInfo));
                }

                tileCollection.AddRange(nestedTiles);

                // Тайл группы.
                var grouptile = new Tile(
                    tileInfo.Caption,
                    TileHelper.SplitCaption(tileInfo.Caption),
                    icon,
                    contextSource,
                    DelegateCommand.Empty,
                    order: tileInfo.Order,
                    group: TileGroups.Workflow,
                    toolTip: tileInfo.Tooltip,
                    tiles: tileCollection,
                    size: tileInfo.TileSize,
                    sharedInfo: CreateInfo(tileInfo));
                if (tileInfo.ActionGrouping)
                {
                    grouptile.SetActionsGrouping(tileInfo.ActionGrouping);
                    grouptile.Evaluating += TileGroupingEvaluation;
                }
                return grouptile;
            }
            var tile = new Tile(
                tileInfo.Name,
                TileHelper.SplitCaption(tileInfo.Caption),
                icon,
                contextSource,
                new DelegateCommand(this.OnClickAsync),
                order: tileInfo.Order,
                group: TileGroups.Workflow,
                toolTip: TileHelper.GetToolTip(tileInfo.Tooltip, tileInfo.ButtonHotkey),
                tiles: tileCollection,
                size: tileInfo.TileSize,
                sharedInfo: CreateInfo(tileInfo));
            if (tileInfo.ActionGrouping)
            {
                tile.SetActionsGrouping(tileInfo.ActionGrouping);
                tile.Evaluating += TileGroupingEvaluation;
            }

            return tile;
        }

        private async void OnClickAsync(object tileObj)
        {
            KrTileInfo tileInfo;

            if (tileObj is ITile tile
                && (tileInfo = tile.SharedInfo.TryGet<KrTileInfo>(KrConstants.Ui.TileInfo)) != null)
            {
                var context = tile.Context;
                var command = tileInfo.IsGlobal
                    ? this.globalCommand
                    : this.localCommand;

                await command.OnClickAsync(context, tile, tileInfo);
            }
        }

        private static SerializableObject CreateInfo(
            KrTileInfo tileInfo)
        {
            return new SerializableObject
            {
                [KrConstants.Ui.TileInfo] = tileInfo
            };
        }

        private static void TileGroupingEvaluation(object sender, TileEvaluationEventArgs e)
        {
            e.SetIsEnabledWithCollapsing(e.CurrentTile, true);
        }
    }
}